template/node/src/service.rs

```diff
@@ -43,7 +46,44 @@ pub enum ConsensusResult {
 	ManualSeal(FrontierBlockImport<Block, Arc<FullClient>, FullClient>)
 }
 
-pub fn new_partial(config: &Configuration, manual_seal: bool) -> Result<
+pub const INHERENT_IDENTIFIER: InherentIdentifier = *b"timstap0";
+
+/// Provide a mock duration since unix epoch in millisecond for timestamp inherent.
+/// First call will return current timestamp and others calls with increment
+/// it by slot_duration to make Aura think time has passed
+pub struct MockInherentDataProvider;
+
+static mut CURRENT_TIMESTAMP: u64 = 0;
+
+impl ProvideInherentData for MockInherentDataProvider {
+	fn inherent_identifier(&self) -> &'static InherentIdentifier {
+		&INHERENT_IDENTIFIER
+	}
+
+	fn provide_inherent_data(
+		&self,
+		inherent_data: &mut InherentData,
+	) -> Result<(), sp_inherents::Error> {
+		unsafe {
+			if CURRENT_TIMESTAMP == 0 {
+				let now = SystemTime::now()
+					.duration_since(SystemTime::UNIX_EPOCH)
+					.unwrap()
+					.as_millis() as u64;
+				CURRENT_TIMESTAMP = now;
+			} else {
+				CURRENT_TIMESTAMP += SLOT_DURATION;
+			}
+			inherent_data.put_data(INHERENT_IDENTIFIER, &CURRENT_TIMESTAMP)
```

Comment:
Is it important that the first timestamp will be close to today's date? Given that the future timestamps will all be forged anyway, why not just start at timestamp 0?

Reply:
yes that will work as well. it will be better for testing because we can easily predict the timestamp for each block

Reply:
@sorpaas @JoshOrndorff making it start from 0